src/error.rs

```diff
@@ -6,17 +6,12 @@ use crate::protocol::{frame::coding::Data, Message};
 use http::Response;
 use thiserror::Error;
 
-#[cfg(feature = "tls")]
-pub mod tls {
-    //! TLS error wrapper module, feature-gated.
-    pub use native_tls::Error;
-}
-
 /// Result type of all Tungstenite library calls.
 pub type Result<T> = result::Result<T, Error>;
 
 /// Possible WebSocket errors.
 #[derive(Error, Debug)]
+#[non_exhaustive]
 pub enum Error {
```

Comment:
While the `#[non_exhaustive]` technically resolves the API incompatibility problems with additivity of features that @sdroege mentioned, I'm not sure whether it is worth it.

It forces downstream crates to add `_ => handle unknown error`, which would be triggered when the crate only handles the "other" Tls error type - which would be most likely unwanted.

It seems that a "proper" solution would to provide a common abstraction of Tls errors. Inside, it could be #[non_exhaustive] enum of possible feature-gated types.

What do you think? Worth it? Note that we already had the API changing based on the `tls` feature flag.

Reply:
You mean to make the error opaque by hiding the real errors behind it and just have a single TLS error?
Or do you mean more of detailing out the error into as many details as each TLS error from the crates provide so we can convert to our own TLS error implementation but still give the user the opportunity to match on the exact TLS problem?

Reply:
I see I was vague in my original comment, so I'd better express that with code:
```rust
/// Possible WebSocket errors. [NOTE: exhaustive]
#[derive(Error, Debug)]
pub enum Error {
    (...)
    Io(#[from] io::Error),
    /// TLS error. [NOTE: enabled unconditionally even if no tls feature is enabled for feature-agnostic API surface]
    #[error("TLS error: {0}")]
    Tls(#[from] TlsError),
    /// - When reading: buffer capacity exhausted.
    (...)
}

/// TLS errors.
///
/// Note that even if you enable only the rustls-based TLS support, the error at runtime could still
/// be `Native`, as another crate in the dependency graph may enable native TLS support.
#[derive(Error, Debug)]
#[non_exhaustive]
pub enum TlsError {
    /// Native TLS error.
    #[cfg(feature = "use-native-tls")]
    #[error("native-tls error: {0}")]
    Native(#[from] native_tls::Error),
    /// Rustls error.
    #[cfg(feature = "use-rustls")]
    #[error("rustls error: {0}")]
    Rustls(#[from] rustls::TLSError),
    /// DNS name resolution error.
    #[cfg(feature = "use-rustls")]
    #[error("Invalid DNS name: {0}")]
    Dns(#[from] webpki::InvalidDNSNameError),
}
```

I've checked what abstraction we provide on top of https://docs.rs/rustls/0.19.0/rustls/enum.TLSError.html and https://docs.rs/native-tls/0.2.7/native_tls/struct.Error.html, but native_tls::Error doesn't provide anything more than `Debug + Display + Error`.

Note that this is a brainstorm rather than a serious suggestion. It depends on what you (and possibly @application-developer-DA, @jxs, @sdroege) think -- whether this is worth it or not.

Reply:
Thanks for all the effort preparing a code sample.

I totally understand now what you mean. That's a good idea to keep the `non_exhaustive` usage limited.

I'll prepare the changes together with your other suggestions tomorrow.